cv32e40s/tests/programs/custom/pma/pma.c

```diff
@@ -13,319 +13,1147 @@
+// status, thus we are limited to 31 tests with this construct.
+#define NUM_TESTS 9
+// Start at 1 (ignore dummy test that is only used for env sanity checking during dev.)
+#define START_TEST_NUM 1
+// Abort test at first self-check fail, useful for debugging.
+#define ABORT_ON_ERROR_IMMEDIATE 0
+
+// Addresses of VP interrupt control registers
+#define TIMER_REG_ADDR         ((volatile uint32_t * volatile) (CV_VP_INTR_TIMER_BASE))
+#define TIMER_VAL_ADDR         ((volatile uint32_t * volatile) (CV_VP_INTR_TIMER_BASE + 4))
+#define DEBUG_REQ_CONTROL_REG *((volatile uint32_t * volatile) (CV_VP_DEBUG_CONTROL_BASE))
+
+// __FUNCTION__ is C99 and newer, -Wpedantic flags a warning that
+// this is not ISO C, thus we wrap this instatiation in a macro
+// ignoring this GCC warning to avoid a long list of warnings during
+// compilation.
+#define SET_FUNC_INFO \
+  _Pragma("GCC diagnostic push") \
+  _Pragma("GCC diagnostic ignored \"-Wpedantic\"") \
+  const volatile char * const volatile name = __FUNCTION__; \
+  _Pragma("GCC diagnostic pop")
+
+// ---------------------------------------------------------------
+// Type definitions
+// ---------------------------------------------------------------
+
+
+// Verbosity levels (Akin to the uvm verbosity concept)
+typedef enum {
+  V_OFF    = 0,
+  V_LOW    = 1,
+  V_MEDIUM = 2,
+  V_HIGH   = 3,
+  V_DEBUG  = 4
+} verbosity_t;
+
+typedef enum {
+  MCAUSE_ACCESS_FAULT    = 1,
+  MCAUSE_ILLEGAL         = 2,
+  MCAUSE_BREAKPT         = 3,
+  MCAUSE_LOAD_FAULT      = 5,
+  MCAUSE_STORE_FAULT     = 7,
+  MCAUSE_UMODE_ECALL     = 8,
+  MCAUSE_MMODE_ECALL     = 11,
+  MCAUSE_INSTR_BUS_FAULT = 24,
+  MCAUSE_CHK_FAULT       = 25,
+} mcause_exception_status_t;
+
+typedef union {
+  struct {
+    volatile uint32_t exccode   : 12;
+    volatile uint32_t res_30_12 : 19;
+    volatile uint32_t interrupt : 1;
+  } __attribute__((packed)) volatile clint;
+  struct {
+    volatile uint32_t exccode    : 12;
+    volatile uint32_t res_15_12  : 4;
+    volatile uint32_t mpil       : 8;
+    volatile uint32_t res_26_24  : 3;
+    volatile uint32_t mpie       : 1;
+    volatile uint32_t mpp        : 2;
+    volatile uint32_t minhv      : 1;
+    volatile uint32_t interrupt  : 1;
+  } __attribute__((packed)) volatile clic;
+  volatile uint32_t raw : 32;
+} __attribute__((packed)) mcause_t;
+
+// ---------------------------------------------------------------
+// Global variables
+// ---------------------------------------------------------------
+
+// Print verbosity, consider implementing this as a virtual
+// peripheral setting to be controlled from UVM.
+volatile verbosity_t global_verbosity = V_LOW;
+
+// Global pointer variables
+volatile uint32_t * volatile g_has_clic;
+volatile uint32_t * volatile g_exp_fault;
+volatile mcause_t * volatile g_mcause;
+volatile uint32_t * volatile g_mepc;
+volatile uint32_t * volatile g_mtval;
+volatile uint32_t * volatile g_test_num;
+volatile uint32_t * volatile g_recovery_ptr;
```

Comment:
Why are these pointers instead of just variables?

Reply:
Convenience, I find working with pointers easier in assembly - could also be variables.